Makes `.mdformat.toml` take precedence

kdeldycke opened the pull request:
This PR closes https://github.com/csala/mdformat-pyproject/issues/17

Additionally, fixes some docstring rST formatting.

kdeldycke commented:
Also, forgot to mentionned, but this PR fixes an issue with mdformat 1.0.0.

The patched `read_toml_opts()` from previously released `mdformat-pyproject` 0.7.x can return inconsistent `(pyproject_opts, pyproject_path)` pair, triggering this traceback:

```pytb
Traceback (most recent call last):
  File "/home/runner/.local/bin/mdformat", line 10, in <module>
    sys.exit(run())
             ^^^^^
  File "/home/runner/.local/share/uv/tools/mdformat/lib/python3.12/site-packages/mdformat/__main__.py", line 8, in run
    exit_code = mdformat._cli.run(sys.argv[1:])
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/.local/share/uv/tools/mdformat/lib/python3.12/site-packages/mdformat/_cli.py", line 56, in run
    opts = {**DEFAULT_OPTS, **toml_opts, **cli_core_opts}
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not a mapping
```

See my CI for more details: https://github.com/kdeldycke/workflows/actions/runs/18754525944/job/53502912176#step:7:20

csala commented on mdformat_pyproject/plugin.py:
```diff
@@ -28,57 +28,92 @@
-        pyproject_opts = _parse_pyproject(pyproject_path)
-    else:
-        pyproject_opts = {}
+    config_file = _find_pyproject_toml_path(conf_dir)
```
I would rename this function to `_find_config_file`, since it's no longer just about `pyproject_toml` (though, I added more comments and this may be removed)

kdeldycke replied:
OK I just removed `_find_pyproject_toml_path` altogether and merged it into `patched_read_toml_opts`.

csala commented on mdformat_pyproject/plugin.py:
```diff
@@ -28,57 +28,92 @@
-        content = tomllib.load(pyproject_file)
+        try:
+            content = tomllib.load(pyproject_file)
+        except tomllib.TOMLDecodeError as e:
```
nice

csala commented on mdformat_pyproject/plugin.py:
```diff
@@ -28,57 +28,92 @@
+                # If we found a pyproject.toml, only return it if it contains
+                # a [tool.mdformat] section.
+                if candidate.name == "pyproject.toml":
+                    options = _parse_pyproject(candidate)
```
Very good point: If pyproject has no tool.mdformat, we ignore it, so an .mdformat.toml in a higher folder takes precendence.

But the current has tow small problems:
- we are parsing the pyproject twice if found and it has options
- only one file is considered at a time, so we issue no warning if both exist and have options.

I suggest changing this a bit as follows:
- Remove this function and move the loop directly to the `read_toml_opts` (since the nested function call makes things too complex in this case)
- Always parse both files (if found) at each level of the loop
- If only one of them exists return its options and path (but ignoring a pyproject.toml without mdtool section)
- If both exist and pyproject has mdformat section, issue a warning and then return only the `.mdformat.toml` options.

kdeldycke replied:
I just implemented all your suggestions in the PR.

csala reviewed (changes requested):
I added some comments. I believe that the current approach may lead to silent conflicts that may be confusing to the user (user edits options in pyproject.toml but observes no behavior changes because a `.mdformat.toml` exists that they did not spot at first)

kdeldycke committed:
Makes `.mdformat.toml` takes precedence

Closes #17

kdeldycke commented:
I addressed all the issues but there are conflicting reports of the pre-commits hooks because of `tests/pre-commit-test.md`. Do you have a preferred way of fixing this?

kdeldycke commented:
> I addressed all the issues but there are conflicting reports of the pre-commits hooks because of `tests/pre-commit-test.md`. Do you have a preferred way of fixing this?

OK I decided to aligns the configuration files to prevent any mismatch.

csala reviewed (changes requested):
@kdeldycke Thank you very much!

There is one last thing: the tests now are a bit redundant, because `test_search_config_directory_inside_project` is almost the same as `test_patched_read_toml_opts_with_pyproject`, and same with the `without` equivalents.

On the other hand, because `.mformat.toml` exists, now there is no test in which the options are actually loaded from the `pyproject.toml` file.

To make it right, I would remove the `test_search_config_**` tests and keep only the `test_patched_read_toml_opts**` versions, but with all these scenarios (possibly using a tempdir to set things up):
- Only `.mdformat.toml` &rarr; Options from `.mdformat.toml` are returned
- Only `pyproject.toml` with options &rarr; No options are returned
- Only `pyproject.toml` without options &rarr; No options are returned
- Both `.mformat.toml` and `pyproject.toml`, but `pyproject.toml` has no options &rarr; Options from `.mdformat.toml` are returned
- Both `.mformat.toml` and `pyproject.toml` exist and with different options &rarr; There is a warning (`mock.patch` the `print_paragraphs` function to certify it) and the options from  `.mdformat.toml` are returned.

I'm happy to make these changes myself, either on top of the same PR or after merging it, if you want, or you can give it a go. Your call.

In any case, I'll make the release with the rest to unblock the compatibility problem, and we can re-release afterwards with this one.

kdeldycke commented:
I can make the test changes, but not today. 😅 So if you are in a hurry do not hesitate to add them on top of this PR!

csala commented:
> I can make the test changes, but not today. 😅 So if you are in a hurry do not hesitate to add them on top of this PR!

Alright, I'll do the changes myself and release a new version with this possibly today.

csala committed:
Improve tests to cover all scenarios

csala commented on mdformat_pyproject/plugin.py:
```diff
@@ -23,63 +24,94 @@
 
 # Monkey patch mdformat._conf to use our own read_toml_opts version
-mdformat._conf.read_toml_opts = read_toml_opts
+mdformat._cli.read_toml_opts = patched_read_toml_opts
```
This is critical! It turns out that we were patching the wrong `read_toml_opts`, and the current plugin version is actually not reading the options correctly!

kdeldycke replied:
Oh good catch!

csala committed:
Use 2 separate files to test pyproject.toml and .mdproject.toml via pre-commit

csala committed:
Fix pre-commit-test regex

kdeldycke commented:
Thanks @csala for the additional tests and merge!

csala commented:
> Thanks @csala for the additional tests and merge!

Thank YOU for moving this in the first place! Good collaboration altogether.
